Feature/59473 Extend Reminders API V3 to include create, update & delete operations

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/wp/59473

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

- [x] `GET /api/v3/reminders`
- [x] `POST /api/v3/work_packages/:work_package_id/reminders`
- [x] `PATCH /api/v3/reminders/:id` 
- [x] `DELETE /api/v3/reminders/:id`
- [x] Add OpenAPI Specification

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
[#59473] Set Reminders API

https://community.openproject.org/work_packages/59473

akabiru committed:
Structure reminder GET utility

akabiru committed:
Add `POST /api/v3/work_packages/:work_package_id/reminders`

akabiru committed:
Add `PATCH /api/v3/work_packages/:work_package_id/reminders/:reminder_id`

akabiru committed:
Add `DELETE /api/v3/work_packages/:work_package_id/reminders/:reminder_id`

akabiru committed:
Add openapi specification for reminders API

akabiru committed:
Update reminders get endpoint to reuse response examples

NobodysNightmare commented on docs/api/apiv3/paths/work_package_reminders.yml:
```diff
@@ -1,61 +1,139 @@
+    - Work Packages
+    - Reminders
+  description: |-
+    Gets a list of your upcoming reminders for this work package.
```
I've read the newly added error message before: Is it true that the "list" of reminders will either be empty or contain exactly one reminder, but never two?

One additional thought: How useful would it be to list reminders independent of a work package? I.e. `/api/v3/reminders`. This could be a way to find _all_ reminders I have set.

akabiru replied:
> I've read the newly added error message before: Is it true that the "list" of reminders will either be empty or contain exactly one reminder, but never two?

Yes, this is true. Only one active reminder is allowed at the moment. Let me touch up the wording to better reflect that. 👍🏾 

> One additional thought: How useful would it be to list reminders independent of a work package? I.e. `/api/v3/reminders`. This could be a way to find _all_ reminders I have set.

This would be useful, and it's something we have planned for the future but out of scope of the current work.

NobodysNightmare replied:
> it's something we have planned for the future but out of scope of the current work.

Is there already a work package for the future work? Looking at the work package linked, I just saw that "reading" own reminders is a requirement, but there was neither out-of-scope mentioned, nor a work package for the future work.

Let's make sure it's written down, if we plan on extending it this way :)

akabiru replied:
Good idea to write down, right now it's an ongoing discussion. I'll take a stab at supporting `/api/v3/reminders` resource in this PR. Will otherwise create a separate wp for it. :)

akabiru replied:
Implemented separately: https://github.com/opf/openproject/pull/19083

NobodysNightmare commented on docs/api/apiv3/paths/work_package_reminders.yml:
```diff
@@ -1,61 +1,132 @@
+            $ref: '../components/schemas/reminder_model.yml'
+    '400':
+      description: |-
+        Returned if the user already has an active reminder for this work package.
```
Did you consider returning a `409 Conflict` for that error case? Sounds like a legitimate (and more specific) error code for this case:

> The 409 (Conflict) status code indicates that the request could not be completed due to a conflict with the current state of the target resource

akabiru replied:
Good call! 👍🏾

akabiru replied:
https://github.com/opf/openproject/pull/19063/commits/4b1f4c9729cbd3b344a41d6d01b54733fe6d38ae

akabiru committed:
Change multiple reminders restriction to HTTP 409

> The 409 (Conflict) status code indicates that the request could not be completed due to a conflict with the current state of the target resource

NobodysNightmare commented:
Two more things that I noticed while using the API myself:

1. There is no `_links/self` included for the reminder itself (this would be useful to have, so that I don't have to construct a reference to it myself)
2. The canonical URL of a reminder always seems to include a work package (because that's where we mounted it), even though the WP id is not necessary to uniquely identify the reminder

The second one is making me question whether we should write the API differently. One way out would be to simplify URLs for reminders, e.g. instead of `DELETE /api/v3/work_packages/:work_package_id/reminders/:id` use `DELETE /api/v3/reminders/:id`.

Alternatively it would seem to be consistent to use `DELETE /api/v3/work_packages/:work_package_id/reminder`, because there can only be one reminder per work package, so it should be clear which reminder you want to delete.

edit: From #19083 I take it that a reminder may be related to something that's not a work package conceptually. In that case I strongly lean towards simpler canonical URLs for reminders (`/api/v3/reminders/:id`) where operations like `DELETE` and `PATCH` are targeted. Because then we only need to define those routes once and the only thing we have to add when attaching reminders to a new container is `/api/v3/<container-name>/:id/reminders`.

akabiru commented:
> edit: From #19083 I take it that a reminder may be related to something that's not a work package conceptually. In that case I strongly lean towards simpler canonical URLs for reminders (`/api/v3/reminders/:id`) where operations like `DELETE` and `PATCH` are targeted. Because then we only need to define those routes once and the only thing we have to add when attaching reminders to a new container is `/api/v3/<container-name>/:id/reminders`.

Cheers, I do like the idea of reducing future amendments 👍🏾 I'll add the change over in #19083

NobodysNightmare commented:
What do you think about **only** reviewing https://github.com/opf/openproject/pull/19083 as a whole diff? I see that feature freeze is approaching and I _hope_ that we can get that PR through until then.

However, what I'd like to avoid is merging **this** PR today and then having breaking changes to the API one release later (moving DELETE from the WP API to the generic spot would for example be a breaking change).

akabiru commented:
> What do you think about **only** reviewing #19083 as a whole diff? I see that feature freeze is approaching and I _hope_ that we can get that PR through until then.
> 
> However, what I'd like to avoid is merging **this** PR today and then having breaking changes to the API one release later (moving DELETE from the WP API to the generic spot would for example be a breaking change).

Sure, that's fine by me. Tbc, I separated them mainly for the benefit of code review- i.e. changes proposed here. We should be able to make it before freeze. 🤞🏾  Addressing your feedback shortly!

akabiru committed:
Add `GET /api/v3/reminders` endpoint (#19083)

An endpoint that lists all reminders for a user, regardless of the workpackage.

akabiru committed:
Support remindable visible? check

akabiru committed:
Simplify reminder PATCH,DELETE urls

"A reminder may be related to something that's not a work package conceptually.
 Lean towards simpler canonical URLs for reminders (`/api/v3/reminders/:id`) where operations like `DELETE` and `PATCH` are targeted.
 Because then we only need to define those routes once and the only thing we have to add when attaching reminders to a new container is `/api/v3/<container-name>/:id/reminders`."
   ~ https://github.com/opf/openproject/pull/19063#issuecomment-2935288456

NobodysNightmare commented on lib/api/v3/reminders/reminder_representer.rb:
```diff
@@ -33,6 +33,27 @@ class ReminderRepresenter < ::API::Decorators::Single
         include API::Decorators::DateProperty
         include API::Decorators::LinkedResource
 
+        def self.associated_remindable_getter
```
This is still lacking a link to self. That's a link we commonly include and we even have a helper for it available.

It should be sufficient to add `self_link` somewhere in the representer definition.

akabiru replied:
On it! 😄

akabiru replied:
Yes, forgot that from your earlier feedback. Added! https://github.com/opf/openproject/pull/19063/commits/3e8a8b637233bce89ff8b8128266dcc8bf5109d9

akabiru committed:
Add missing self link for reminder representation

NobodysNightmare reviewed (approved):
LGTM. Thanks for the extensions.
